Fixes #1763 - Flaky test: SimpleTimerTest

DomGarguilo opened the pull request:
Did my best to resolve #1763. Replaced the assert statements with an if statement so that they could be repeated if they don't pass immediately. The test will repeat until passed or until the added timeout occurs. As suggested, I left the assert statement after the loop which verifies the thread pool size remains unchanged after the previous operations. Not sure if I should have kept the other asserts as well but I don't see the point if the same conditions were verified already. Please suggest any changes or improvements.

ctubbsii commented on server/base/src/test/java/org/apache/accumulo/server/util/time/SimpleTimerTest.java:
```diff
@@ -98,11 +97,15 @@ public void testFailure() throws InterruptedException {
-    assertSame(t, t2);
+  @Test(timeout = 5000)
+  public void testSingleton() throws InterruptedException {
+    while (true) {
+      SimpleTimer t2 = SimpleTimer.getInstance(2);
+      if((SimpleTimer.getInstanceThreadPoolSize() == 1) && (t == t2)) {
+        break;
+      }
+      Thread.sleep(PAD);
+    }
```
This is the right strategy (albeit with a few extra parenthesis than necessary on that `if` statement :wink: )

DomGarguilo replied:
Oops, I see I was off-base with this one. Will do my best to correct these issues with your suggestions in mind :+1:

ctubbsii reviewed (changes requested):
Hi @DomGarguilo . Thanks for the PR! You have the correct strategy in place. However, the test case that you modified was probably the only test case for this class that isn't prone to the type of failure that was reported. It should pass every time. If you look at the other 3 tests, you can see that they all employ a "sleep" already, in order to try to catch the test in the correct state. However, since that's flaky (it may take slightly longer for the task to start), those other 3 tests are the ones that should employ the method you applied here (instead of this one).

EDIT: also, please make sure the changed files are formatted to our coding standards by running `mvn package -DskipTests` at the root of the project, and including any changes made by our formatting plugins to your commits / PR.

DomGarguilo commented:
Oops, I see I was off-base with this one. Will do my best to correct these issues with your suggestions in mind +1

DomGarguilo commented:
I implemented the fix in the correct test cases. There were a few things that I was unsure about though. Not sure if I should extract the code containing the while(true) and create a separate method in order to reuse that portion of code. Not sure if I should add a local variable for the timeout length (even though there is one test case that requires a longer timeout). I added a timeout to testSingleton() which probably doesn't need it so I am unsure whether I should remove it or keep it in.

ctubbsii commented on server/base/src/test/java/org/apache/accumulo/server/util/time/SimpleTimerTest.java:
```diff
@@ -70,35 +69,50 @@ public void run() {
     t.schedule(r, DELAY);
     Thread.sleep(DELAY + PAD);
-    assertEquals(1, i.get());
+    while (true) {
+      if (i.get() == 1) {
+        break;
+      }
+      Thread.sleep(PAD);
+    }
```
Just realized that you can simplify this to `while (i.get() != 1) { sleep }`; you can do similar things for the other tests as well.

DomGarguilo replied:
Was considering that as well. I will add this change.

ctubbsii commented:
> Not sure if I should extract the code containing the while(true) and create a separate method in order to reuse that portion of code.

No. These changes aren't big enough to warrant that.

> Not sure if I should add a local variable for the timeout length (even though there is one test case that requires a longer timeout).

No. Your timeouts are fine.

> I added a timeout to testSingleton() which probably doesn't need it so I am unsure whether I should remove it or keep it in.

In general, we try to make minimal unrelated changes, to make it easier to review, but it's often a judgement call. In this case, I would leave it in. It's a tiny change, and is related to your adding a timeout to all the other test cases.

ctubbsii commented:
Thanks for improving the test, @DomGarguilo !
